netlib-lapack: Fixes for IBM XL builds

quellyn opened the pull request:
This PR addresses #25447, build failures encountered with `netlib-lapack@3.9.1%xl_r`. It incorporates some of the fixes provided in an earlier patch, `ibm-xl.patch` (the rest have been upstreamed).

Additionally, it adds correct detection for the `xlf`'s `-qrecur` option. Note that the current logic will test ALL possible recursive flags against a target compiler; in the case of `xlf` this unexpectedly "succeeds" when the Intel-specific `-recursive` flag is passed in. Actually it doesn't really succeed, but `xlf` does not correctly report failure, so `cmake` CheckFortranCompilerFlag ends up tacking `-recursive` onto `CMAKE_Fortran_FLAGS`. The resulting attempt to compile with this bogus flag fails.

You can see the errant compiler behavior with a simple example "Hello World" program:
```
$ xlf_r -recursive -o hello-dumb hello.f90
** helloworld   === End of Compilation 1 ===
1501-510  Compilation successful for file hello.f90.
ld: cannot find -lgcc_s
ld: cannot find -lgcc_s

$ ls
hello.f90  hello.o

$ file hello.o
hello.o: ELF 64-bit LSB relocatable, 64-bit PowerPC or cisco 7500, version 1 (SYSV), not stripped
```
The `xlf` compiler produces an intermediate object code that will not link.

I've reported this problem to the [LAPACK maintainers](https://github.com/Reference-LAPACK/lapack/issues/606#issuecomment-888463053) and will direct them here so they can evaluate this solution for upstreaming.

Commit:
hypre: Add releases 2.21.0 and 2.22.0

Commit:
Revert "hypre: Add releases 2.21.0 and 2.22.0"

This reverts commit 8921cdb3ac772794f82ff07b0aa816c204aca921.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' of github.com:quellyn/spack into develop

Commit:
Address external linkage failures in elfutils 0.185:
  https://bugs.gentoo.org/794601
  https://sourceware.org/pipermail/elfutils-devel/2021q2/003862.html
Encountered while building within a Spack environment.

Commit:
Merge branch 'develop' of github.com:quellyn/spack into develop

Commit:
Revert "Address external linkage failures in elfutils 0.185:"

This reverts commit 76b93e4504defd856b62027371064a5e65a027a5.

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
Merge branch 'develop' of github.com:quellyn/spack into develop

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Commit:
netlib-lapack: Version 3.9.0 and above no longer builds with the IBM XL
compiler (#25447). Ported some fixes from the old ibm-xl.patch and added
logic for detection of XL's -qrecur flag.

sethrj commented on var/spack/repos/builtin/packages/netlib-lapack/package.py:
```diff
@@ -42,10 +42,13 @@ class NetlibLapack(CMakePackage):
 
-    patch('ibm-xl.patch', when='@3.7: %xl')
-    patch('ibm-xl.patch', when='@3.7: %xl_r')
+    patch('ibm-xl.patch', when='@3.7:3.8.9999 %xl')
+    patch('ibm-xl.patch', when='@3.7:3.8.9999 %xl_r')
```
Spack version semantics are quite tricky, so it's hard to remember that the above is equivalent to the simpler and less arbitrary:
```suggestion
    patch('ibm-xl.patch', when='@3.7:3.8 %xl')
    patch('ibm-xl.patch', when='@3.7:3.8 %xl_r')
```

To test the version specification (thanks to @tgamblin for teaching me this trick):
```console
$ spack-python
Spack version 0.16.2
Python 3.8.5, Darwin x86_64
>>> from spack.version import ver
>>> ver('@3.8.999') in ver('@3.7:3.8')
```

quellyn replied:
> Spack version semantics are quite tricky, so it's hard to remember that the above is equivalent to the simpler and less arbitrary:
> 
> To test the version specification (thanks to @tgamblin for teaching me this trick):
> 
> ```
> $ spack-python
> Spack version 0.16.2
> Python 3.8.5, Darwin x86_64
> >>> from spack.version import ver
> >>> ver('@3.8.999') in ver('@3.7:3.8')
> ```

I'm happy to use the simpler syntax. I only used the "four-nines" syntax because it's used elsewhere in the `netlib-lapack` recipe.

sethrj reviewed (commented):
The new cmake looks good to me, though I'm unable to test the updated logic.

Commit:
Update var/spack/repos/builtin/packages/netlib-lapack/package.py

I'm happy to adopt the shorter syntax. I only used the "four-nines" syntax because it's used elsewhere in the `netlib-lapack` recipe.

Co-authored-by: Seth R. Johnson <johnsonsr@ornl.gov>
